gcsio/src/main/java/com/google/cloud/hadoop/gcsio/CreateObjectOptions.java

```diff
@@ -19,126 +19,81 @@
- */
-public class CreateObjectOptions {
-  public static final ImmutableMap<String, byte[]> EMPTY_METADATA = ImmutableMap.of();
-  public static final String DEFAULT_CONTENT_TYPE = "application/octet-stream";
-  public static final String DEFAULT_CONTENT_ENCODING = null;
-  public static final CreateObjectOptions DEFAULT =
-      new CreateObjectOptions(/* overwriteExisting= */ true);
-
-  private final boolean overwriteExisting;
-  private final String contentType;
-  private final String contentEncoding;
-  private final Map<String, byte[]> metadata;
-  private final boolean requireMetadataMatchForEmptyObjects;
+/** Options that can be specified when creating a file in the {@link GoogleCloudStorage}. */
+@AutoValue
+public abstract class CreateObjectOptions {
 
-  /**
-   * Construct a new CreateObjectOptions with empty metadata and the default content-type.
-   * Since no metadata was explicitly provided, we will set metadata-match for
-   * empty objects to false.
-   *
-   * @param overwriteExisting True to overwrite any existing objects with the same name.
-   */
-  public CreateObjectOptions(boolean overwriteExisting) {
-    this(overwriteExisting, DEFAULT_CONTENT_TYPE, DEFAULT_CONTENT_ENCODING, EMPTY_METADATA, false);
-  }
+  public static final String CONTENT_TYPE_DEFAULT = "application/octet-stream";
 
-  /**
-   * Construct a new CreateObjectOptions with the specified metadata, and default content-type.
-   * Since metadata was explicitly provided, even if empty, we will set metadata-match for
-   * empty objects to true.
-   * 
-   *
-   * @param overwriteExisting True to overwrite any existing objects with the same name.
-   * @param metadata A dictionary of metadata to apply to created objects.
-   */
-  public CreateObjectOptions(boolean overwriteExisting, Map<String, byte[]> metadata) {
-    this(overwriteExisting, DEFAULT_CONTENT_TYPE, metadata);
-  }
+  public static final CreateObjectOptions DEFAULT_NO_OVERWRITE = builder().build();
+  public static final CreateObjectOptions DEFAULT_OVERWRITE =
+      builder().setOverwriteExisting(true).build();
 
-  /**
-   * Construct a new CreateObjectOptions with the spec metadata and content-type.
-   * Since metadata was explicitly provided, even if empty, we will set metadata-match for
-   * empty objects to true.
-   *
-   * @param overwriteExisting True to overwrite any existing objects with the same name
-   * @param contentType content-type for the created file
-   * @param metadata A dictionary of metadata to apply to created objects
-   */
-  public CreateObjectOptions(
-      boolean overwriteExisting, String contentType, Map<String, byte[]> metadata) {
-    this(overwriteExisting, contentType, DEFAULT_CONTENT_ENCODING, metadata, true);
+  public static Builder builder() {
+    return new AutoValue_CreateObjectOptions.Builder()
+        .setContentEncoding(null)
+        .setContentType(CONTENT_TYPE_DEFAULT)
+        .setEnsureEmptyObjectsMetadataMatch(true)
```

Comment:
Shouldn't this default to be false? So that later you create default for empty object you don't need to set it back to false?

Reply:
Yeah, I checked how this change came about, and it seems that it's safer to set it `true` because it will guarantee that created object has the same metadata that we need.